db/db_impl/db_impl_open.cc

```diff
@@ -1158,266 +1169,456 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
   // No-op for immutable_db_options_.wal_filter == nullptr.
   InvokeWalFilterIfNeededOnColumnFamilyToWalNumberMap();
 
+  *min_wal_number = MinLogNumberToKeep();
+  if (!allow_2pc()) {
+    // In non-2pc mode, we skip WALs that do not back unflushed data.
+    *min_wal_number =
+        std::max(*min_wal_number, versions_->MinLogNumberWithUnflushedData());
+  }
+}
+
+Status DBImpl::ProcessLogFiles(
+    const std::vector<uint64_t>& wal_numbers, bool read_only, bool is_retry,
+    uint64_t min_wal_number, int job_id, SequenceNumber* next_sequence,
+    std::unordered_map<int, VersionEdit>* version_edits,
+    bool* corrupted_wal_found, RecoveryContext* recovery_ctx) {
+  Status status;
+
   bool stop_replay_by_wal_filter = false;
   bool stop_replay_for_corruption = false;
   bool flushed = false;
   uint64_t corrupted_wal_number = kMaxSequenceNumber;
-  uint64_t min_wal_number = MinLogNumberToKeep();
-  if (!allow_2pc()) {
-    // In non-2pc mode, we skip WALs that do not back unflushed data.
-    min_wal_number =
-        std::max(min_wal_number, versions_->MinLogNumberWithUnflushedData());
-  }
+
   for (auto wal_number : wal_numbers) {
-    if (wal_number < min_wal_number) {
-      ROCKS_LOG_INFO(immutable_db_options_.info_log,
-                     "Skipping log #%" PRIu64
-                     " since it is older than min log to keep #%" PRIu64,
-                     wal_number, min_wal_number);
-      continue;
+    if (status.ok()) {
+      status =
+          ProcessLogFile(wal_number, min_wal_number, is_retry, read_only,
+                         job_id, next_sequence, &stop_replay_for_corruption,
+                         &stop_replay_by_wal_filter, &corrupted_wal_number,
+                         corrupted_wal_found, version_edits, &flushed);
+    }
+  }
+
+  if (status.ok()) {
+    status = MaybeHandleStopReplayForCorruptionForInconsistency(
+        stop_replay_for_corruption, corrupted_wal_number);
+  }
+
+  if (status.ok()) {
+    status = MaybeFlushFinalMemtableOrRestoreActiveLogFiles(
+        wal_numbers, read_only, job_id, flushed, version_edits, recovery_ctx);
+  }
+  return status;
+}
+
+Status DBImpl::ProcessLogFile(
+    uint64_t wal_number, uint64_t min_wal_number, bool is_retry, bool read_only,
+    int job_id, SequenceNumber* next_sequence, bool* stop_replay_for_corruption,
+    bool* stop_replay_by_wal_filter, uint64_t* corrupted_wal_number,
+    bool* corrupted_wal_found,
+    std::unordered_map<int, VersionEdit>* version_edits, bool* flushed) {
+  assert(stop_replay_by_wal_filter);
+
+  // Variable initialization starts
+  Status status;
+  bool old_log_record = false;
```

Comment:
I moved old_log_record to be local to each log file instead of across log files. This should be fine cuz `HandleNonOkStatusOrOldLogRecord()` resets the old_log_record in the case that matters (non- WALRecoveryMode::kSkipAnyCorruptedRecords recovery modes).